### PR TITLE
Clean up HString & add exception to Locate

### PR DESCRIPTION
**Pull request by ericpoe**
HString is full of methods that allow it to behave like an \ArrayObject. This PR extracts those methods out to an abstract class (abstract, because I don't want it to be instantiated) that does those things for it.

While I was at it, I also made the `::locate(x)` method throw an ElementNotFoundException instead of returning a "-1" since it's possible for a "-1" to be an appropriate key. Sigh. This should have been a separate branch, but I was feeling all-powerful at the time.

Since the change in the `::locate(x)` method causes a BC break, this will require that Haystack gets a major revision change once merged in.

### CODE REVIEW

**Comment by johnkary** [comment]
### BaseHString

What are the benefits to introducing a base HString class? Would BaseHString be extended in multiple ways, either by this library or in a user's codebase?

I'm hesitant to introduce yet another class to separate the HString's custom API from the ArrayObject API it mimics because introducing BaseHString increases the complexity of the code without an equal or greater benefit. Inheritance has been introduced to HString and it goes from having no parent class to now having one.
1. To reason about HString the developer must keep in mind a lot of behavior from a second file/class BaseHString.
2. By introducing inheritance the developer must **decide** which class they want to extend (or compose)... HString or BaseHString? We can embrace the "[Don't Make Me Think](https://en.wikipedia.org/wiki/Don%27t_Make_Me_Think)" mindset by removing the need for a decision and only having one extension point: HString.
3. Inheritance should only be used when its concrete implementations are true "is a" relationships. Inheritance is discouraged when used to simply facilitate code reuse. So far this looks like code separation/reuse. Currently Haystack only offers one implementation of BaseHString... HString.

A side note... Using the logic for introducing BaseHString to segment the ArrayObject API, shouldn't it follow that BaseHArray also be created and it contain the overridden methods from ArrayObject, then HArray extends BaseHArray with custom API? The added class in the hierarchy doesn't add much value.

### `locate()` now throwing an Exception

I like that `locate()` now throws an Exception if not located. If a developer is unsure a given array contains an element they should use `contains() : bool`. Then using `'locate()` when they know it's in there--and it not being in there is Exceptional. Some devs will probably dislike this idea because core PHP itself is less strict. "Why not just check the return type for false?!" but because PHP can use so many things as array keys throwing an Exception is more clear and makes better sense.

Re: major version bump, you wouldn't need to immediately tag a new major version. A library's master branch can be bleeding edge. But since this API change breaks BC it would be time to signal that the master branch is now 2.0-dev. See [Composer dev-alias](https://getcomposer.org/doc/articles/aliases.md) and [an example](https://github.com/symfony/symfony/blob/master/composer.json#L108) from the Symfony codebase.

I believe this is the dev-alias you would want to add to this PR to signal 2.0 development has begun:

``` json
    "extra": {
        "branch-alias": {
            "dev-master": "2.0-dev"
        }
    }
```

Then more BC changes could be added to master before a new 2.0 version is tagged. Features we've kicked around like making `HArray::map()` variadic and adding different `split*()` methods.

**Comment by ericpoe** [comment]
@johnkary I appreciate your feedback. You are, of course, absolutely right about the use of abstract. I'll work on these changes over the next few days.

File: src/Functional/HStringFilter.php [hunk 1]
```diff
@@ -6,35 +6,35 @@
 class HStringFilter
 {
     /** @var HString */
-    protected $string;
+    protected $hSTring;
```
Comment (johnkary) [inline src/Functional/HStringFilter.php:9]: Small typo: `hSTring` should probably be `hString`

Renaming protected class properties is considered a BC break. In this instance any class extending HStringFilter and using `$this->string` would require updating their code to use `$this->hSTring` as the new property name. (But I don't think this class is likely to be extended in any other user's code so it doesn't need documenting this time :D )

**Comment by ericpoe** [comment]
Doh! Good catch. Updated and documented.

File: src/Functional/HStringFilter.php (hunk 1 above)
Comment (ericpoe) [inline src/Functional/HStringFilter.php:9]: Doh! Good catch. Updated and documented.

**Pull request merged**